Combine log entries (Prereq for OC-1541)

bradenmacdonald opened the pull request:
This change combines `GeneralLogEntry`, `InstanceLogEntry`, `ServerLogEntry` into a single `LogEntry` class that uses [the contenttypes framework](https://docs.djangoproject.com/en/1.9/ref/contrib/contenttypes/) to hold a reference to the object that "owns" each log entry.

This change is required in order to accommodate the refactoring in #63. Specifically, there is a new intermediate class, `AppServer`, which would need a new table to hold its logs. And `AppServer` is abstract (since multi-table inheritance in django gets very messy), so there are potentially a number of concrete `AppServer` subclasses that we may have, and we can't expect to create a separate logging table for each one. Luckily, django already has a wonderful and longstanding solution to this problem, which is the contenttypes framework.

Pros:
* Code is way simpler
* Loading logs way faster/simpler
* Less load on app server
* Ignoring the migrations, this change resulting in a net reduction of 77 lines of code :)
* ~~Log entries not deleted when instance/server is deleted.~~
* Includes a fully reversible data migration, so this can be tested, deployed, or even reverted without any loss of data.
* The various existing logging test cases cover this code and pass with (almost) no modifications.

Cons:
* Foreign keys not enforced at DB level.
* ~~(As a result:) Log entries not deleted when instance/server is deleted. (Not sure if this is a pro or a con)~~

Also, while I was at it I fixed an occasional pylint warning about undocumented Meta classes: We don't need docstrings on Meta, and now pylint knows that, without inline disable comments

antoviaque commented:
@bradenmacdonald Cool, I didn't knew about content types, this looks like a good solution. Thanks for doing the change!

>Cons:
>
>    Foreign keys not enforced at DB level.
>    (As a result:) Log entries not deleted when instance/server is deleted. (Not sure if this is a pro or a con)

This sounds like things you can handle with signals, no? Ie pre_save & pre_delete? Would be good to not leave objects behind when the instances/servers are deleted, especially. 

Or for foreign keys is it already enforced at the application level?

bradenmacdonald commented:
Yes, exactly. I can add a post_delete signal to delete related log entries when a server or instance is deleted. I notice that we weren't deleting the general logs ever so I thought we might also want to preserve the server/instance logs forever too, and maybe delete them after a year or so rather than when the server is deleted. But I'm not sure that would actually be useful, so I will go ahead with a post_delete signal.

bradenmacdonald commented:
I've now added code to delete old log entries automatically.

antoviaque commented:
@bradenmacdonald For the general log entries, they should have been deleted with the instance itself, no? If not that would have been a bug. Generally speaking we'll probably want to avoid deleting instance/server objects in any case (and flag them as deleted instead), but it's always good to make sure that if we do we don't keep orphaned objects; the database could end up growing to a point where that becomes necessary, especially for the logentries.

Btw my remark about signals was also about your other point, "_Foreign keys not enforced at DB level._" - even if it can't be enforced at the DB level, `pre_save` could help to enforce this at the application level, if it's not already?

bradenmacdonald commented:
@antoviaque As far as I know, the general log is full of entries like "200 GET /" ... but maybe that's only when run in debug mode. Those general log entries are usually not associated with a particular instance, so would never be deleted.

Foreign keys are not strictly enforced at the application level, but the code as it stands would never create an invalid foreign key. I don't think there's any reason to add a separate validation layer.

antoviaque commented:
@bradenmacdonald You shouldn't be reading your email on your birthday :p 

For these general log entries, it could be good having a closer look - are they generated by a provision run? Could be worth adding a bug in the backlog if that's the case, and if that's not addressed by the current PR.

> Foreign keys are not strictly enforced at the application level, but the code as it stands would never create an invalid foreign key. I don't think there's any reason to add a separate validation layer.

Today yes (or maybe? How can you be sure there isn't a bug?) - but an explicit validation of this in pre_save would ensure it remains the case, and doesn't take much to put in place.

itsjeyd commented on instance/migrations/0047_generic_logging3.py:
```diff
@@ -0,0 +1,45 @@
+    def data_backward(apps, schema_editor):
+        """
+        Reverse the data migration
+        """
```
@bradenmacdonald I think you can remove `data_forward` and `data_backward` from this migration.

bradenmacdonald replied:
Whoops, good point - thanks.

itsjeyd commented on instance/migrations/0047_generic_logging3.py:
```diff
@@ -0,0 +1,45 @@
+        migrations.RemoveField(
+            model_name='serverlogentry',
+            name='obj',
+        ),
```
@bradenmacdonald Why do we need to remove these fields before deleting the model?

bradenmacdonald replied:
That was auto-generated by Django's migration framework. I believe the reason is that these are foreign keys, and some databases require you to delete the foreign key constraints before deleting the table.

itsjeyd commented on instance/models/instance.py:
```diff
@@ -268,65 +270,27 @@ def save(self, **kwargs):
         if limit:
-            return entries[-limit:]
-        return entries
+            return reversed(list(entries[:limit]))
```
@bradenmacdonald Nit: When I first read this I was tempted to suggest that you could do `entries[:limit].order_by('created')` here, to be consistent with the code in the next line. But that doesn't work because "Cannot reorder a query once a slice has been taken." Maybe add a comment to that extent, to avoid potential confusion?

bradenmacdonald replied:
@itsjeyd  In SQL, there's no simple way to say "Sort the rows by increasing date and return the last 200 rows"; all you can say is "sort the rows by *decreasing* date and return up to 200 rows". So since `order_by` directives must be translated to SQL, it makes sense that Django can't do that.


I will add a comment that says "# Apply the limit at the SQL/DB level while sorted by descending date, then reverse. Otherwise, we'd have to retrieve all rows and then apply the limit using python."

itsjeyd commented on instance/models/log_entry.py:
```diff
@@ -52,41 +50,38 @@ class LogEntry(ValidateModelMixin, TimeStampedModel):
+            return  # Current LogEntry schema requires integer objects IDs, so we know this object isn't in the table
+        content_type = ContentType.objects.get_for_model(instance)
+        LogEntry.objects.filter(content_type=content_type, object_id=instance.pk).delete()
+        # TODO: Use the following instead once we upgrade to Django 1.9:
```
@bradenmacdonald Leftover TODO, just marking it to ensure we don't forget

bradenmacdonald replied:
@itsjeyd Django 1.9 was merged so I updated this PR with master and uncommented this.

itsjeyd commented on instance/tests/models/test_log_entry.py:
```diff
@@ -133,6 +133,36 @@ def test_log_publish(self, mock_publish_data):
+        self.assertEqual(LogEntry.objects.count(), 2)
+        entries = LogEntry.objects.order_by('pk').all().values_list('text', flat=True)
+        self.assertIn('Line #1, on instance', entries[0])
+        self.assertIn('Line #3, on server 2', entries[1])
```
@bradenmacdonald Nit: Maybe add another check asserting that "Line #2, on server 1" is missing from `entries[1]`.

itsjeyd commented on instance/models/log_entry.py:
```diff
@@ -21,24 +21,22 @@
 """
 
 # Imports #####################################################################
-
```
Nit: We should keep this blank line to stay consistent with the formatting in other files.

itsjeyd commented on instance/models/log_entry.py:
```diff
@@ -21,24 +21,22 @@
 
 logger = logging.getLogger(__name__)
 
-
```
@bradenmacdonald Nit: We should keep this blank link to stay consistent with the formatting in other files.

itsjeyd commented:
@bradenmacdonald Happy birthday! :)

I reviewed this and left a few comments. Manual testing went fine. I was able to 

* run forward and backward migrations 
* get rid of log entries by deleting corresponding server and instance objects
* provision a new instance

without issue. Running the migrations produced expected results (in both directions).

bradenmacdonald commented:
@antoviaque @itsjeyd All review comments addressed. Is this good to merge now?

itsjeyd commented on instance/tests/models/test_log_entry.py:
```diff
@@ -134,6 +136,108 @@ def test_log_publish(self, mock_publish_data):
+
+    def test_invalid_content_type_object_id_combo(self):
+        """
+        Test that content_type cannot be set on its own
```
@bradenmacdonald Nit: "Test that content_type and object_id cannot be set on their own"

I'm fine with updating this docstring as part of #63; it's definitely not a blocker to merging this PR :)

itsjeyd commented:
@bradenmacdonald Thanks for the updates, this LGTM :+1:

antoviaque commented:
@bradenmacdonald Yup, LGTM - I have only had a look at the changes we discussed, but I don't have any remaining objections.

bradenmacdonald commented:
Thanks guys!

Addressed nit, rebased from 473371f7cfa1d07d8c09d68e0f75aa2c184d0f39 onto master... will merge once green.

Rebased diff is identical, except for the nit I fixed.
